src/Router/index.tsx

```diff
@@ -1,123 +1,67 @@
-import React, { useState, useEffect, useRef } from 'react';
+import React, { useState, useEffect, useRef, useCallback } from 'react';
 import { createBrowserHistory, History } from 'history';
 import { RouterContext } from '../Context';
 
-import { Route } from '../types';
+import { Route, Router as IRouter } from '../types';
 import Path from '../PathUtils';
 
-export interface RouterState {
-  location: string;
-  params: Object;
-  routes: Array<Route>;
-  context: Object;
-  forceRefresh: number;
-  routedElement: React.ReactNode | undefined;
-}
-
 export interface IRouterProps {
   routes: Array<Route>;
   children: React.ReactNode
 }
 
-export const Router: React.FC<IRouterProps> = props => {
-  const history = useRef<History | undefined>(undefined);
-  const initialState: RouterState = {
-    location: '',
-    params: {},
-    context: {},
-    routes: props.routes,
-    routedElement: undefined,
-    forceRefresh: 0
-  };
+const browserHistory = createBrowserHistory()
 
-  const [state, setState] = useState(initialState);
+const getComponentFromRoute = (route: Route, router: IRouter) => {
+  let guardReturn: React.ReactNode = undefined;
 
-  const setLocation = (location: string) => {
-    if (state.location !== location) {
-      history.current?.push(location);
-    } else {
-      setState({
-        ...state,
-        forceRefresh: state.forceRefresh + 1
-      });
+  if (route.guards) {
+    let nexted = false;
+    for (const guard of route.guards) {
+      guardReturn = guard(router, () => {
+        nexted = true
+        return undefined
+      })
+      if (!nexted && guardReturn) break
     }
-  };
-
-  const setContext = (context: Object) => {
-    setState({
-      ...state,
-      context: Object.assign(state.context, context)
-    });
-  };
+  }
 
-  useEffect(() => {
-    history.current = createBrowserHistory();
-    setState({
-      ...state,
-      location: history.current.location.pathname
-    });
-
-    const unlisten = history.current.listen(update => {
-      setState({
-        ...state,
-        location: update.location.pathname
-      });
-    });
+  return guardReturn ?? route.component
+}
 
-    return () => {
-      unlisten();
-    };
-  }, []);
+export const Router = ({ routes, children }: IRouterProps) => {
+  const [location, setLocation] = useState<string>(browserHistory.location.pathname)
+  const [context, setContext] = useState<object>({})
```

Comment:
I minimized the state necessary to run the router.

Reply:
Nice! With this every change in context will trigger the router logic. The only concern I have is if it could be useful to change the router context without triggering a reroute?